### PR TITLE
chacha20rng: add AVX backend

### PR DESCRIPTION
**Pull request by ripatel-fd**

### CODE REVIEW

File: src/ballet/chacha20/fd_chacha20_avx.c [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+  wu_t k2 = _mm256_shuffle_epi32( key_lo, 0xaa );
+  wu_t k3 = _mm256_shuffle_epi32( key_lo, 0xff );
+  wu_t k4 = _mm256_shuffle_epi32( key_hi, 0x00 );
+  wu_t k5 = _mm256_shuffle_epi32( key_hi, 0x55 );
+  wu_t k6 = _mm256_shuffle_epi32( key_hi, 0xaa );
+  wu_t k7 = _mm256_shuffle_epi32( key_hi, 0xff );
+
+  /* Derive block index */
+
+  ulong idx = rng->buf_fill / FD_CHACHA20_BLOCK_SZ;  /* really a left shift */
```
Comment (ptaffet-jump) [inline src/ballet/chacha20/fd_chacha20_avx.c:50]:
```suggestion
  ulong idx = rng->buf_fill / FD_CHACHA20_BLOCK_SZ;  /* really a right shift */
```

File: src/ballet/chacha20/fd_chacha20_avx.c [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  cC = wu_add( cC, idxs );
+  //cD = wu_add( cD, zero );
+  //cE = wu_add( cE, zero );
+  //cF = wu_add( cF, zero );
+
+  /* Transpose matrix to get output vector */
+
+  /* Note: transpose_vecs originally derives from the BLAKE3 C official
+           implementation.  See NOTICE for licensing details. */
+#define transpose_vecs( h0, h1, h2, h3, h4, h5, h6, h7 ) do {          \
```
Comment (ptaffet-jump) [inline src/ballet/chacha20/fd_chacha20_avx.c:103]: Is this the same as `wu_transpose_8x8`?

File: src/ballet/chacha20/fd_chacha20_avx.c [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+  wu_st( out+0x20, c2 ); wu_st( out+0x28, cA );
+  wu_st( out+0x30, c3 ); wu_st( out+0x38, cB );
+  wu_st( out+0x40, c4 ); wu_st( out+0x48, cC );
+  wu_st( out+0x50, c5 ); wu_st( out+0x58, cD );
+  wu_st( out+0x60, c6 ); wu_st( out+0x68, cE );
+  wu_st( out+0x70, c7 ); wu_st( out+0x78, cF );
+
+  /* Update ring descriptor */
+
+  rng->buf_fill += (uint)( 8*FD_CHACHA20_BLOCK_SZ );
```
Comment (ptaffet-jump) [inline src/ballet/chacha20/fd_chacha20_avx.c:154]: You don't need the cast anymore, right?

**Review by ptaffet-jump — DISMISSED** [review]

**Review by ptaffet-jump — APPROVED** [review]

**Review by ptaffet-jump — DISMISSED** [review]
I just noticed the tests are failing

**Comment by ripatel-fd** [comment]
@ptaffet-jump Oops, sorry. Fixed that.

**Review by mmcgee-jump — APPROVED** [review]

**Pull request merged**